nemo/collections/asr/parts/mixins/transcription.py

```diff
@@ -0,0 +1,542 @@
+
+        try:
+            generator = self.transcribe_generator(audio, override_config=transcribe_cfg)
+
+            for processed_outputs in generator:
+                # Store results
+                if isinstance(processed_outputs, list):
+                    # Create a results of the same type as each element in processed_outputs
+                    if results is None:
+                        results = []
+
+                        # if list of inner list of results, copy structure
+                        if isinstance(processed_outputs[0], list):
+                            for _ in processed_outputs:
+                                results.append([])
+
+                    # If nested list structure
+                    if isinstance(processed_outputs[0], list):
+                        for i, processed_output in enumerate(processed_outputs):
+                            results[i].extend(processed_output)
+                    else:
+                        # If flat list structure
+                        results.extend(processed_outputs)
+
+                elif isinstance(processed_outputs, dict):
+                    # Create a results of the same type as each element in processed_outputs
+                    if results is None:
+                        results = processed_outputs
+                    else:
+                        for k, v in processed_outputs.items():
+                            results[k].extend(v)
+
+                elif isinstance(processed_outputs, tuple):
+                    # Create a results of the same type as each element in processed_outputs
+                    if results is None:
+                        results = tuple([[] for _ in processed_outputs])
+
+                    # If nested list structure
+                    if isinstance(processed_outputs[0], list):
+                        for i, processed_output in enumerate(processed_outputs):
+                            results[i].extend(processed_output)
+                    else:
+                        # If flat list structure
+                        if len(processed_outputs) != len(results):
+                            raise RuntimeError(
+                                f"The number of elements in the result ({len(results)}) does not "
+                                f"match the results of the current batch ({len(processed_outputs)})."
+                            )
+
+                        for i, processed_output in enumerate(processed_outputs):
+                            results[i].append(processed_output)
+
+                else:
+                    raise NotImplementedError(
+                        "Given output result for transcription is not supported. "
+                        "Please return a list of results, list of list of results, "
+                        "a dict of list of results, or "
+                        "a tuple of list of results."
+                    )
+        except StopIteration:
```

Comment:
I can't see where `transcribe_generator` would raise a `StopIteration`. Internally it uses a `for` loop over the dataloader which wouldn't raise that. Unless I'm missing something this generator would just gracefully finish at the end of iteration.

Reply:
Apparently in newer python versions, return in a generator counts as a StopIteration. I could be explicit but it's supported since Py 3.8 so eh

Reply:
right, and that raised `StopIteration` is implicitly caught by `for` statement in l. 222: `for processed_outputs in generator:`, isn't it?

Reply:
Added an explicit test, seems to work fine. It raises a StopIteration after consuming all samples in the input

Reply:
What I wanted to communicate is that `try/except StopIteration` shouldn't be needed here because the `for` loop already handles StopIteration. See e.g. the test that you wrote `test_transcribe_generator` and compare with `test_transcribe_generator_expliit_stop_check` (btw there's a typo in the name). When you don't use a `for` loop (explicit_stop_check) you need to explicitly catch it. When you use a `for` loop it won't raise that. But if this is the only comment blocking you from merging then don't mind it and go ahead.